### ISSUE TITLE
Wire Reactivity for Mod Music Boxes

### ISSUE THREAD

**Issue by TecmagDiams** [bluemagic123/tModLoader#3182]
### Do you intend to personally contribute/program this feature?

No

### I would like to see this change made to improve my experience with

Gameplay as a Player

### Description

Currently tModLoader has a function to register Music Boxs however it does not include the vanilla behavior that makes them reactive to wiring requiring each mod to implement it's own wire behavior. All major mods I have tested do not appear to individually add this behavior which makes them not act 1:1 like the vanilla music boxes in-game.

Rather than require each modder to create this (although simple) behavior it feels more correct that tModLoader's call to register a music box should automatically create this behavior so that all mod music boxes are consistent with vanilla music boxes.

### What does this proposal attempt to solve or improve?

This proposal attempts to solve inconsistencies between vanilla and modded music box behavior. Music boxes are already a commonly used feature inside mods and mod developers should not need to individually replicate a vanilla behavior on each music box.

### Which (other) solutions should be considered?

The alternative is leaving this behavior up to modders, which creates inconsistencies with vanilla behavior.

**Comment by Kogsey** [comment]
This had already been done and shouldn't be an issue, in tModLoader's Wiring patch it has:

```
case139:
case int _ when TileLoader.IsModMusicBox(tile):
    WorldGen.SwitchMB(i, j);
    break;
```

So the case for normal Music Box (139) falls into the tModLoader and they both follow the same logic, I used copy-pasted example mod code the check this, no manual wiring needed. I would assume from this that any changes in music box behaviour from mods are intentional?

Edit: Thought I should also mention the same function is used to modify the vanilla Music Box player interactions

**Comment by TecmagDiams** [comment]
Thanks for looking into this! Taking what you've said I've done some additional testing to try and limit down what might cause this issue (and possibly make it more complex and lower priority).

I've identified this issue specifically happens in multiplayer. When started singleplayer I am able to flip a switch to toggle music boxes, but loading the same world in "Host & Play" or as a dedicated server results in the mod music boxes not reacting to wires (where vanilla ones still do). I've attached a quick example of the issue below.

(For those unfamiliar who shouldn't need to ask: the music box on the left from Verdant should open/close in it's states, and the Mushroom from Calamity should light up and go back. I had each one in an opposite state for a more comprehensive test)

https://user-images.githubusercontent.com/48166478/211872427-3bc29fd9-5ec4-4d7b-b4ff-bf14f21b5c84.mp4

**Comment by Kogsey** [comment]
Well I found the issue, the server side MusicLoader.tileToMusic is empty because AddMusicBox is called in SetStaticDefaults, a Client side method, this leads the server side wire code spitting the dummy in TileLoader.IsModMusicBox, meaning no return in the wire switch-case. meaning no action, I've been called into work early so I'll tinker for a reasonable solution tommorow, just figured I'd leave this here just in case it saves someone some trouble

**Comment by JavidPack** [comment]
SetStaticDefaults isn't client side, the issue is AddMusicBox will return early on the server, we'll have to think more about this, but it seems fixable.

**Comment by Chicken-Bones** [comment]
Fixed in #3191
